### PR TITLE
PR AB#79500 document service for portal

### PR DESCRIPTION
**Pull request by pawelpwc**

### CODE REVIEW

**Review by dawid-pawulski — APPROVED** [review]

File: src/HE.InvestmentLoans.WWW/Controllers/CompanyStructureV2Controller.cs [hunk 1]
```diff
@@ -65,36 +65,76 @@ public async Task<IActionResult> PurposePost(Guid id, CompanyStructureViewModel
...
+        return File(response.Data, "application/octet-stream", response.Name);
+    }
+
+    [HttpPost("more-information-about-organization")]
+    [WorkflowState(CompanyStructureState.ExistingCompany)]
+    public async Task<IActionResult> MoreInformationAboutOrganizationPost(Guid id, CompanyStructureViewModel viewModel, [FromForm(Name = "File")] IFormFile formFile, CancellationToken cancellationToken)
+    {
+        if (formFile != null)
+        {
+            var result = await _mediator.Send(
```
Comment (Carq) [inline src/HE.InvestmentLoans.WWW/Controllers/CompanyStructureV2Controller.cs:106]: There should be only one command which saves organization information and upload files.
Upload should happen only after clicking "Continue" button"

File: src/HE.InvestmentLoans.BusinessLogic/CompanyStructure/CommandHandlers/ProvideMoreInformationAboutOrganizationFileCommandHandler.cs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    IRequestHandler<ProvideMoreInformationAboutOrganizationFileCommand, OperationResult>
+{
+    private readonly IDocumentServiceConfig _config;
+
+    private readonly IHttpDocumentService _documentService;
+
+    public ProvideMoreInformationAboutOrganizationFileCommandHandler(
+                ICompanyStructureRepository repository,
+                ILoanUserContext loanUserContext,
+                IAppConfig appConfig,
```
Comment (Carq) [inline src/HE.InvestmentLoans.BusinessLogic/CompanyStructure/CommandHandlers/ProvideMoreInformationAboutOrganizationFileCommandHandler.cs:30]: AppConfig is not needed.

File: src/HE.InvestmentLoans.WWW/wwwroot/javascripts/ExistingCompany.file-validation.js [hunk 3]
```diff
@@ -1,25 +0,0 @@
-function validateFileSize() {
```
Comment (Carq) [inline src/HE.InvestmentLoans.WWW/wwwroot/javascripts/ExistingCompany.file-validation.js:1]: Please Revert the change. File should be validation on client site before uploading to the server.

File: src/HE.InvestmentLoans.Contract/CompanyStructure/ValueObjects/OrganisationMoreInformationFiles.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+using HE.InvestmentLoans.Common.Utils.Constants;
+using HE.InvestmentLoans.Common.Validation;
+using HE.InvestmentLoans.Contract.Domain;
+using HE.Investments.DocumentService.Models.File;
+using MediatR;
+using Microsoft.AspNetCore.Http;
+
+namespace HE.InvestmentLoans.Contract.CompanyStructure.ValueObjects;
+
+public class OrganisationMoreInformationFiles : ValueObject
```
Comment (Carq) [inline src/HE.InvestmentLoans.Contract/CompanyStructure/ValueObjects/OrganisationMoreInformationFiles.cs:12]: Imho. There should be one Value Object which aggregates all files and validate all files. We can discuss this offline.

File: src/HE.InvestmentLoans.Contract/CompanyStructure/ValueObjects/OrganisationMoreInformationFile.cs [hunk 5]
```diff
@@ -17,33 +20,29 @@ public class OrganisationMoreInformationFile : ValueObject
         AllowedFileExtension.RTF,
     };
 
-    public OrganisationMoreInformationFile(string fileName, byte[] content, int maxFileSizeInMb)
+    public OrganisationMoreInformationFile(FileData file, int maxFileSizeInMb)
```
Comment (Carq) [inline src/HE.InvestmentLoans.Contract/CompanyStructure/ValueObjects/OrganisationMoreInformationFile.cs:23]: PLease revert of this change. Classes from external service cannot be part Domain classes (ValueObjects, entities)

File: src/HE.InvestmentLoans.BusinessLogic/CompanyStructure/QueryHandlers/GetCompanyStructureFilesQueryHandler.cs [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    {
+        _documentService = documentService;
+        _config = config;
+    }
+
+    public async Task<TableResult<FileTableRow>> Handle(GetCompanyStructureFilesQuery request, CancellationToken cancellationToken)
+    {
+        var folderPath = "0000000_DA2123DAE440EE11BDF3002248C653E1";
+
+        var result = await _documentService.GetTableRowsAsync(new FileTableFilter()
```
Comment (Carq) [inline src/HE.InvestmentLoans.BusinessLogic/CompanyStructure/QueryHandlers/GetCompanyStructureFilesQueryHandler.cs:30]: All communication with Document service (upload, download, remove etc)  should be part of the Data layer so it should be in repositories.

File: src/HE.InvestmentLoans.BusinessLogic/CompanyStructure/CommandHandlers/CompanyStructureBaseCommandHandler.cs [hunk 7]
```diff
@@ -23,14 +24,14 @@ public CompanyStructureBaseCommandHandler(ICompanyStructureRepository repository
         _logger = logger;
     }
 
-    protected async Task<OperationResult> Perform(Action<CompanyStructureEntity> action, LoanApplicationId loanApplicationId, CancellationToken cancellationToken)
+    protected async Task<OperationResult> Perform(Func<CompanyStructureEntity, UserAccount, Task> action, LoanApplicationId loanApplicationId, CancellationToken cancellationToken)
```
Comment (Carq) [inline src/HE.InvestmentLoans.BusinessLogic/CompanyStructure/CommandHandlers/CompanyStructureBaseCommandHandler.cs:27]: This is needed only in one command handler, so you don't have use base command handler, just create own one and please revert that change.

**Review by Carq — CHANGES_REQUESTED** [review]

File: src/HE.InvestmentLoans.BusinessLogic/CompanyStructure/CompanyStructureEntity.cs [hunk 8]
```diff
@@ -33,6 +31,8 @@ public class CompanyStructureEntity
 
     public OrganisationMoreInformationFile? MoreInformationFile { get; private set; }
 
+    public OrganisationMoreInformationFiles? MoreInformationFiles { get; private set; }
```
Comment (Carq) [inline src/HE.InvestmentLoans.BusinessLogic/CompanyStructure/CompanyStructureEntity.cs:34]: It should be set via ctor

**Review by Carq — CHANGES_REQUESTED** [review]

File: src/HE.InvestmentLoans.BusinessLogic/CompanyStructure/CommandHandlers/ProvideMoreInformationAboutOrganizationFileCommandHandler.cs [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+                companyStructure.ProvideFileWithMoreInformation(new OrganisationMoreInformationFile(file, _config.MaxFileSizeInMegabytes));
+
+                await _documentService.UploadAsync(new FileUploadModel()
+                {
+                    ListTitle = _config.ListTitle,
+                    FolderPath = "0000000_DA2123DAE440EE11BDF3002248C653E1",
+                    File = file,
+                    Metadata = JsonSerializer.Serialize(new FileMetadata
+                    {
+                        CreateDate = DateTime.Now,
```
Comment (Carq) [inline src/HE.InvestmentLoans.BusinessLogic/CompanyStructure/CommandHandlers/ProvideMoreInformationAboutOrganizationFileCommandHandler.cs:57]: `IDateTimeProvider.Now` should be used.

File: src/HE.InvestmentLoans.BusinessLogic/CompanyStructure/CommandHandlers/ProvideMoreInformationAboutOrganizationFileCommandHandler.cs [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+                        Creator = userAccount.AccountName,
```
Comment (Carq) [inline src/HE.InvestmentLoans.BusinessLogic/CompanyStructure/CommandHandlers/ProvideMoreInformationAboutOrganizationFileCommandHandler.cs:58]: AccountName is the organisation name. You should put First and Last name of the user.

**Review by Carq — CHANGES_REQUESTED** [review]

File: src/HE.InvestmentLoans.BusinessLogic/CompanyStructure/CommandHandlers/ProvideMoreInformationAboutOrganizationFileCommandHandler.cs (hunk 2 above)
Comment (pawelpwc) [reply src/HE.InvestmentLoans.BusinessLogic/CompanyStructure/CommandHandlers/ProvideMoreInformationAboutOrganizationFileCommandHandler.cs:30]: Is needed

File: src/HE.InvestmentLoans.Contract/CompanyStructure/ValueObjects/OrganisationMoreInformationFiles.cs (hunk 4 above)
Comment (pawelpwc) [reply src/HE.InvestmentLoans.Contract/CompanyStructure/ValueObjects/OrganisationMoreInformationFiles.cs:12]: The class validate all files (files.count)

File: src/HE.InvestmentLoans.Contract/CompanyStructure/ValueObjects/OrganisationMoreInformationFile.cs (hunk 5 above)
Comment (Carq) [reply src/HE.InvestmentLoans.Contract/CompanyStructure/ValueObjects/OrganisationMoreInformationFile.cs:23]: Why it was marked as resolved?

**Review by Carq — APPROVED** [review]

**Pull request merged**